staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -1155,32 +1160,15 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
 		listOptions = listOptions.DeepCopy()
 	}
 
-	listObj, err := e.List(ctx, listOptions)
-	if err != nil {
-		return nil, err
-	}
-	items, err := meta.ExtractList(listObj)
-	if err != nil {
-		return nil, err
-	}
-	if len(items) == 0 {
-		// Nothing to delete, return now
-		return listObj, nil
-	}
-	// Spawn a number of goroutines, so that we can issue requests to storage
-	// in parallel to speed up deletion.
-	// It is proportional to the number of items to delete, up to
-	// DeleteCollectionWorkers (it doesn't make much sense to spawn 16
-	// workers to delete 10 items).
+	var items []runtime.Object
+
+	// TODO(wojtek-t): Decide if we don't want to start workers more opportunistically.
 	workersNumber := e.DeleteCollectionWorkers
-	if workersNumber > len(items) {
-		workersNumber = len(items)
-	}
 	if workersNumber < 1 {
 		workersNumber = 1
 	}
 	wg := sync.WaitGroup{}
-	toProcess := make(chan int, 2*workersNumber)
+	toProcess := make(chan runtime.Object, 2*workersNumber)
```

Comment:
nit: DeleteCollectionWorkers is 1 by default (or small in general), isn't it? So we work through nearly all deletions of a big pages list, and only shortly before the end we list the next page. Maybe we don't care or even want that.

Reply:
I think default is 1, but IIRC most of providers bump it (GKE does), but indeed the values are usually small (I think we do 16 or sth like that).

I would be fine with bumping it a bit, but I'm also reluctant to bump it too high to avoid listing more than we can even process. So for now, I'm bumping it to `max(2*workers, 256)` - I suggest doing that and improving further once we have better evidence.